package fresco.opus.CtlCmdObjects;

enum CC_MSG_TYPE {
    CMDCTL = 1;
    CMDCTLRSP = 2;
    PSRSP = 3;
    QRYRSP = 4;
    STATRSP = 5;
}

enum ServiceStatus {
    LIVE = 1;
    DEAD = 2;
    NOT_PRESENT = 3;
}

message CmdCtlMessage {

    message KVPair {
        optional string key = 1;
        optional string value = 2;
    }

    optional string cmd_name = 1; // opus-ps, opus-kill etc.
    repeated KVPair args = 2; // list of arg name and value
}

message CmdCtlMessageRsp {
    optional string rsp_data = 1; // anything in string format
}

message PSMessageRsp {

    message PSData {
        optional int64 pid = 1; // Process ID
        optional int64 thread_count = 2; // Number of active threads for the process
    }

    repeated PSData ps_data = 1; // Process data
}

message QueryMessageRsp {
    message Row {
        message KVPair {
            optional string key = 1;
            optional string value = 2;
        }

        repeated KVPair cells = 1;
    }

    repeated Row rows = 1;
    optional string error = 2;
}

message StatusMessageRsp {

    message AnalyserStatus {
        optional ServiceStatus status = 1;
        optional int64 num_msgs = 2;
    }

    optional AnalyserStatus analyser_status = 1;
    optional ServiceStatus producer_status = 2;
    optional ServiceStatus query_status = 3;
}
